### PR TITLE
New Check License for Readme and Plugin Header

### LINKED ISSUE #436: Check: GPL Compatible license declared
The plugins must be GPL compatible, so we have to warn to developers to fix this issue.

It has to be declared in readme and plugin. And we have to check if they are the same in both files.

This is the text that we are using to warn users.

```
## No GPL-compatible license declared

It is necessary to declare the license of this plugin. You can do this using the fields available both in the plugin readme and in the plugin headers.

Remember that all code, data, and images — anything stored in the plugin directory hosted on WordPress.org — must comply with the GPL or a GPL-Compatible license. Included third-party libraries, code, images, or otherwise, must be also compatible

For a specific list of compatible licenses, [please read the GPL-Compatible license list on gnu.org](https://www.gnu.org/licenses/license-list.html#GPLCompatibleLicenses).

The license declared on this plugin is either no present or it's not GPL compatible:

ERROR: License declaration not found on tiqq-shop.php
Please check the header requirements in order to correctly declare the license of the plugin: https://developer.wordpress.org/plugins/plugin-basics/header-requirements/
It is necessary to configure the "License: " field in the plugin header with a valid, GPL-compliant license name.
```

### PR DESCRIPTION
**Pull request by davidperezgar**
With this new check solves the issue #436 with this checks:
- Verify that Readme has GPL license.
- Verify that Plugin header has GPL license.
- Verify that is the same license between two files.

Fixes #436

### CODE REVIEW

File: includes/Checker/Default_Check_Repository.php
Comment (mukeshpanchal27) [inline includes/Checker/Default_Check_Repository.php]: Why new check registered? Instead of it can we introduce new method in `Plugin_Readme_Check`

**Review by mukeshpanchal27 — DISMISSED** [review]
Thanks @davidperezgar for the PR. Left one blocker feedback also needs to add unit tests.

File: includes/Checker/Default_Check_Repository.php
Comment (davidperezgar) [reply includes/Checker/Default_Check_Repository.php]: New check is needed becouse it compares readme and plugin header.

Comment (swissspidy) [reply includes/Checker/Default_Check_Repository.php]: We could potentially do that in the existing check too if desired

Comment (davidperezgar) [reply includes/Checker/Default_Check_Repository.php]: I don't see a problem. I've created to use for checks that has to compare plugin header with the readme.

Comment (swissspidy) [reply includes/Checker/Default_Check_Repository.php]: If we combine the two classes there would be less redundancy and maintenance overhead

File: includes/Checker/Checks/Plugin_Readme_Check.php
Comment (frantorres) [inline includes/Checker/Checks/Plugin_Readme_Check.php:255]: Recommended refactor for the following lines.

```
$license = str_replace('-', '', $license);
$license = str_replace('GNU General Public License (GPL)', 'GPL', $license);
$license = str_replace('GNU General Public License', 'GPL', $license);
$license = preg_replace('/GPL\s*[-|\.]*\s*[v]?([0-9])(\.[0])?/i', 'GPL$1', $license, 1);
$license = str_replace('.', '', $license);
```

**Review by frantorres — CHANGES_REQUESTED** [review]
Some refactoring in the code.

**Comment by davidperezgar** [comment]
I've updated my tests for the new check. Please check @Pacomarchante and @frantorres

**Comment by davidperezgar** [comment]
Hey, the testing is not working because this reason:ç

`[2024-06-22T15:14:50.797Z] ['error'] There was an error running the uploader: Error uploading to [https://codecov.io:](https://codecov.io/) Error: There was an error fetching the storage URL during POST: 429 - {'detail': ErrorDetail(string='Rate limit reached. Please upload with the Codecov repository upload token to resolve issue. Expected time to availability: 1674s.', code='throttled')}`

**Comment by swissspidy** [comment]
This is expected at the moment because PRs from forks don't have access to repository secrets like the codecov token. It can be simply ignored. If you create a PR from a branch directly in this repo, this won't happen.

File: includes/Checker/Checks/Plugin_Readme_Check.php
Comment (swissspidy) [inline includes/Checker/Checks/Plugin_Readme_Check.php]: Author attribution like this is not something we currently add in this repo.

**Comment by swissspidy** [comment]
Would be great to add new tests for this change :-)

**Comment by davidperezgar** [comment]
Yeah, This PR I've made before I have acces to the repo. But it's blocking the merge.

**Comment by davidperezgar** [comment]
It has the necessary test for different license. How can we solve the merging problem?

**Comment by swissspidy** [comment]
I just see a lot of code being added but not any new tests. Not a huge deal, but I would have expected more test coverage there.

The failing test does not block anything. Merging is blocked because 2 reviewers requested changes and there are 0 approvals. You can dismiss stale reviews but you still need approvals. Once a PR is approved, it can be merged.

**Comment by davidperezgar** [comment]
Ok!

**Comment by frantorres** [comment]
Looks good. Just a detail.
The new check was named `different_license` and there are already other check with the name `stable_tag_mismatch`, I propose to change it to `license_mismatch` to follow the convention.

**Comment by davidperezgar** [comment]
Done!

**Review by frantorres — APPROVED** [review]

File: includes/Checker/Checks/Plugin_Readme_Check.php
Comment (mukeshpanchal27) [inline includes/Checker/Checks/Plugin_Readme_Check.php:225]: Should add `@Since` for next version

File: includes/Checker/Checks/Plugin_Readme_Check.php
Comment (mukeshpanchal27) [inline includes/Checker/Checks/Plugin_Readme_Check.php]: Can we update function name to `normalice_licenses`?

File: includes/Checker/Checks/Plugin_Readme_Check.php
Comment (mukeshpanchal27) [inline includes/Checker/Checks/Plugin_Readme_Check.php:170]: Why it removed? We should early return if we don't have anything to check.

File: includes/Checker/Checks/Plugin_Readme_Check.php
Comment (mukeshpanchal27) [inline includes/Checker/Checks/Plugin_Readme_Check.php]: Can we use `$license` here as it's duplicate.

**Review by mukeshpanchal27 — COMMENTED** [review]
Thanks @davidperezgar for the updates. Left some feedbacks for your consideration

**Comment by davidperezgar** [comment]
Changes asked by @mukeshpanchal27 done!

**Review by Pacomarchante — APPROVED** [review]
With the latest correction, the tests work correctly.

**Pull request merged**
